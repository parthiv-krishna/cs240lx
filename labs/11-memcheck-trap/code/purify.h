#ifndef __PURIFY_H__
#define __PURIFY_H__

#include "src-loc.h"


void purify_init(void);
#define purify_alloc(_n) purify_alloc_raw(_n, SRC_LOC_MK())
#define purify_free(_ptr) purify_free_raw(_ptr, SRC_LOC_MK())

void *purify_alloc_raw(unsigned n, src_loc_t loc);
void purify_free_raw(void *p, src_loc_t loc);

int purify_heap_errors(void);

// just for the debug macro.
#include "memtrace.h"

static inline void purify_verbose_set(int v_p) {
    memtrace_verbose_set(v_p);
}
#endif
