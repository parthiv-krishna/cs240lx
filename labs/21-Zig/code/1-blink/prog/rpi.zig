extern fn nop() void;

// countdown 'ticks' cycles
pub fn delay(ticks: u32) void {
    var cnt = @as(u32, ticks);
    while (cnt > 0) {
        nop();
        cnt -= @as(u32, 1);
    }
}


