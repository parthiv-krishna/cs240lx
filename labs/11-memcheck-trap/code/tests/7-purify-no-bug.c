// no bug: allocate a couple blocks, memset, free.
#include "rpi.h"
#include "purify.h"

unsigned expected = 45;

static int fn(void) {
    // zero initialized by kmalloc.
    volatile int *x = purify_alloc(4);
    for(int i = 0; i < 10; i++)
        *x += i;
    printk("*x = %d\n", *x);
    assert(get32(x) == expected);
    return *x;
}

void notmain(void) {
    trace("should have no error: running w verbose off\n");
    purify_init();
    purify_verbose_set(0);

    if(fn() != expected)
        panic("invalid result\n");

    trace("SUCCESS: no bug\n");
}
