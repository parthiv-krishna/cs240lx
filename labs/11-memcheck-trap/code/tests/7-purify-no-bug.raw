find-ttyusb.c:find_ttyusb:55:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
pi-install: tty-usb=</dev/ttyUSB0> program=<./tests/7-purify-no-bug.bin> about to boot
HASH: simple_boot: sending 28904 bytes, crc32=a4f85673
waiting for a start
putting code
pi sent print: <<Flynn Dreilinger Lab 4>: success: Received the program!
>
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
TRACE:should have no error: running w verbose off
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x8074]: load from address 0x183ef8
TRACE:[pc=0x807c]: store to address 0x183ef8
TRACE:[pc=0x808c]: load from address 0x183ef8
*x = 45
TRACE:[pc=0x802c]: load from address 0x183ef8
TRACE:[pc=0x80b0]: load from address 0x183ef8
TRACE:SUCCESS: no bug
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
