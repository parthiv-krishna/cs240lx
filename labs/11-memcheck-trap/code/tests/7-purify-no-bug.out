TRACE:should have no error: running w verbose off
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: store to address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:: load from address 0x183ef8
TRACE:SUCCESS: no bug
